keps/sig-storage/20200212-volume-group.md

```diff
@@ -0,0 +1,1067 @@
+
+### Restore
+
+Restore can be done as follows:
+1. A new empty volume group can be created first, and then a new volume can be created from a snapshot one by one and added to the volume group. This can be repeated for all the snapshots in the VolumeGroupSnapshot.
+2. A VolumeGroup can be created from a VolumeGroupSnapshot source in one step. This is the same as what is described in the section `Create VolumeGroup from VolumeGroupSnapshot`.
+
+### API Definitions
+
+API definitions are as follows:
+
+```
+type VolumeGroupClass struct {
+        metav1.TypeMeta
+        // +optional
+        metav1.ObjectMeta
+ 
+        // Driver is the driver expected to handle this VolumeGroupClass.
+        // This value may not be empty.
+        Driver string
+ 
+        // Parameters holds parameters for driver.
+        // These values are opaque to the system and are passed directly
+        // to the driver.
+        // +optional
+        Parameters map[string]string
+
+        // This field specifies whether group snapshot is supported.
+        // The default is false.
+        // +optional
+        VolumeGroupSnapshot *bool
+
+        // Specifies whether consistent group snapshot is supported.
+	// The default is false.
+        // +optional
+        ConsistentGroupSnapshot *bool
+}
+
+// VolumeGroup is a user's request for a group of volumes
+type VolumeGroup struct {
+        metav1.TypeMeta
+        // +optional
+        metav1.ObjectMeta
+
+        // Spec defines the volume group requested by a user
+        Spec VolumeGroupSpec
+
+        // Status represents the current information about a volume group
+        // +optional
+        Status *VolumeGroupStatus
+}
+
+// VolumeGroupSpec describes the common attributes of group storage devices
+// and allows a Source for provider-specific attributes
+Type VolumeGroupSpec struct {
+        // +optional
+        VolumeGroupClassName *string
+
+        // If InitSource is nil, an empty volume group will be created.
+        // Otherwise, a volume group will be created with PVCs.
+	// If Selector is set in InitSource, existing PVCs with matching
+	// label will be added to the volume group.
+	// If SourceVolumeGroupSnapshotName is not nil, the volume group
+	// will be created from the source VolumeGroupSnapshot.
+	// This field determines what PVCs will be in the volume group
+	// when it is initially created. PVCs can be added to or removed
+	// from the volume group later if CSI driver supports
+	// VOLUMEGROUP_ADD_REMOVE_EXISTING_VOLUME.
+        // +optional
+        InitSource *VolumeGroupSource
+}
+
+// VolumeGroupSource contains 2 options. If VolumeGroupSource is not nil,
+// one and only one of the 2 options must be defined.
+Type VolumeGroupSource struct {
+        // A label query over existing persistent volume claims to be added to the volume group.
+        // +optional
+        Selector *metav1.LabelSelector
```

Comment:
Can volume group controller modify PVC spec to add group name?

Reply:
Suggest to add a new CRD and controller to select labeled PVCs.

Reply:
> Can volume group controller modify PVC spec to add group name?
No. PVC spec is modified by a user to add a group name. Volume group controller watches the PVCs and reacts to the update event.